java/arcs/core/util/Scheduler.kt

```diff
@@ -0,0 +1,249 @@
+    /* internal */
+    val launches = atomic(0)
+    /* internal */
+    val loops = atomic(0)
+    private var processingJob: Job? = null
+    private val scope = CoroutineScope(context)
+    private val isIdle = atomic(true)
+    private val isPaused = atomic(false)
+    private val agenda = atomic(Agenda())
+
+    /** Schedule a single [Task] to be run as part of the next agenda. */
+    fun schedule(task: Task) {
+        agenda.update { it.addTask(task) }
+        if (isIdle.value) scope.startProcessingJob()
+    }
+
+    /** Schedule some [Task]s to be run as part of the next agenda. */
+    fun schedule(tasks: Iterable<Task>) {
+        agenda.update { it.addTasks(tasks) }
+        if (isIdle.value) scope.startProcessingJob()
+    }
+
+    /** Pause evaluation of the agenda. */
+    fun pause() {
+        isPaused.value = true
+    }
+
+    /** Resume evaluation of scheduled [Task]s in the agenda. */
+    fun resume() {
+        if (!isPaused.compareAndSet(expect = true, update = false)) return
+        // If we were paused, we can re-start.
+        scope.startProcessingJob()
+    }
+
+    /** Wait for the [Scheduler] to become idle. */
+    suspend fun waitForIdle() {
+        processingJob?.join()
+    }
+
+    private fun CoroutineScope.startProcessingJob() {
+        isIdle.value = false
+        processingJob = launch {
+            launches.incrementAndGet()
+
+            var shallContinue = true
+            while (shallContinue) {
+                val startTime = time.currentTimeMillis
+                shallContinue = withTimeout(agendaProcessingTimeout) { process() }
+                val elapsed = time.currentTimeMillis - startTime
+
+                if (shallContinue) {
+                    loops.incrementAndGet()
+
+                    // If the operation was super fast, let's delay a little bit to allow the
+                    // thread to chill out.
+                    val delayLength = scheduleRateHz.hzToMillisPerIteration() - elapsed
+                    if (delayLength > 0) delay(delayLength)
+                }
+            }
+        }.also { it.invokeOnCompletion { isIdle.value = true } }
+    }
+
+    private suspend fun process(): Boolean = suspendCancellableCoroutine { cont ->
+        val timeoutHandler = { throwable: Throwable ->
+            if (throwable is TimeoutCancellationException) {
+                log.error(throwable) { "Scheduled tasks timed out." }
+            }
+        }
+
+        if (isPaused.value) {
+            cont.resume(false, timeoutHandler)
+            return@suspendCancellableCoroutine
+        }
+
+        val agenda = agenda.getAndSet(Agenda())
+        if (agenda.isEmpty()) {
+            cont.resume(false, timeoutHandler)
+            return@suspendCancellableCoroutine
+        }
+
+        log.debug { "Processing agenda: $agenda" }
+
+        // Process agenda
+        for (task in agenda) {
+            task()
+            if (cont.isCancelled) break
```

Comment:
I am still unclear on what the addition of the `suspendCancellableCoroutine` buys us. Couldn't a simple `yield()` here with regular `suspend` function achieve the same result? (along with a try-catch around the `process` call in the `withTimeout` block?

Reply:
Yield will allow a subsequent launch to start to process an agenda "concurrently" (even if it's single threaded - they'd be woven inbetween eachother, which is not what we want -- so I didn't want to do any actual suspending).

Reply:
ugh..I have more reading to do.